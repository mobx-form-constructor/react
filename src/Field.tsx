import React, { ComponentType, ComponentPropsWithRef, ReactNode } from 'react'
import { Field as IField } from '@mobx-form-constructor/core'
import { useObserver } from 'mobx-react-lite'

export type BindType = 'input' | 'checkbox' | 'radio'

export interface IBaseProps<
  F extends IField,
  T extends BindType,
  C extends FieldComponentType = ComponentType<FieldComponentProps>
> {
  field: F
  type?: T
  value?: F['value']
  component: C
}

export type GenericComponentType = 'input' | 'textarea' | 'select'
export type FieldComponentType = ComponentType<any> | GenericComponentType

export type Bindings<
  F extends IField,
  T extends BindType = 'input'
> = ReturnType<
  T extends 'input'
    ? F['bind']
    : T extends 'checkbox'
    ? F['bindCheckbox']
    : T extends 'radio'
    ? F['bindRadio']
    : never
>

export type FieldComponentProps<
  F extends IField = IField,
  T extends BindType = 'input'
> = Bindings<F, T> & {
  field: F
  error: F['error']
  children: ReactNode
}

export function Field<
  C extends FieldComponentType,
  F extends IField,
  T extends BindType
>({
  field,
  type,
  value,
  component,
  ...rest
}: (C extends ComponentType<infer P>
  ? Omit<P, 'component' | keyof FieldComponentProps> &
      Partial<FieldComponentProps<F, T>>
  : ComponentPropsWithRef<C>) &
  IBaseProps<F, T, C>) {
  return useObserver(() => {
    const bindings =
      type === 'radio'
        ? field.bindRadio(value)
        : type === 'checkbox'
        ? field.bindCheckbox()
        : field.bind()

    const error = (field.touched || field.form.submitFailed) && field.error

    return React.createElement(component, {
      ...bindings,
      ...rest,
      type,
      error: typeof component !== 'string' ? error : undefined,
      field: typeof component !== 'string' ? field : undefined
    })
  })
}
