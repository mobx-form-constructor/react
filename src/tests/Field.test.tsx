import React from 'react'
import { Form } from '@mobx-form-constructor/core'
import { render } from '@testing-library/react'

import { Field, FieldComponentProps } from '../Field'

import { Model } from './Model'

interface ICustomInput extends FieldComponentProps {
  label: string
}

export function CustomInput({
  value,
  onChange,
  onBlur,
  onFocus,
  label, // custom property
  field, // field is forwarded
  error,
  children
}: ICustomInput) {
  return (
    <div>
      <div>key: {field.key}</div>
      <label>
        {label}
        <input
          value={value}
          onChange={onChange}
          onBlur={onBlur}
          onFocus={onFocus}
        />
      </label>
      {error && <span>{error}</span>}
      {children}
    </div>
  )
}

function FieldTestComponent() {
  const form = new Form(Model)
  return (
    <>
      <Field
        field={form.fields.username}
        component={CustomInput}
        label="Username"
      >
        children
      </Field>
      <Field
        field={form.fields.username}
        component={CustomInput}
        label="Username"
      ></Field>

      <Field field={form.fields.username} component="select" placeholder="sd">
        children
      </Field>
      <Field
        field={form.fields.username}
        component="select"
        placeholder="sd"
      ></Field>
    </>
  )
}

describe('Field', () => {
  it('should renders correctly without any type error', () => {
    const { container } = render(<FieldTestComponent />)

    expect(container).toMatchSnapshot()
  })
})
