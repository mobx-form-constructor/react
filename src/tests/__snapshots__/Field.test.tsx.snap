// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field should renders correctly without any type error 1`] = `
<div>
  <div>
    <div>
      key: 
      username
    </div>
    <label>
      Username
      <input
        value=""
      />
    </label>
    children
  </div>
  <div>
    <div>
      key: 
      username
    </div>
    <label>
      Username
      <input
        value=""
      />
    </label>
  </div>
  <select
    placeholder="sd"
  >
    children
  </select>
  <select
    placeholder="sd"
  />
</div>
`;
